extensions/azure/processors/PutAzureBlobStorage.h

```diff
@@ -0,0 +1,138 @@
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <utility>
+#include <vector>
+#include <string>
+#include <memory>
+
+#include "core/Property.h"
+#include "core/Processor.h"
+#include "core/logging/Logger.h"
+#include "core/logging/LoggerConfiguration.h"
+#include "storage/BlobStorage.h"
+#include "utils/OptionalUtils.h"
+
+class PutAzureBlobStorageTestsFixture;
+
+namespace org {
+namespace apache {
+namespace nifi {
+namespace minifi {
+namespace azure {
+namespace processors {
+
+class PutAzureBlobStorage : public core::Processor {
+ public:
+  static constexpr char const* ProcessorName = "PutAzureBlobStorage";
+
+  // Supported Properties
+  static const core::Property ContainerName;
+  static const core::Property AzureStorageCredentialsService;
+  static const core::Property StorageAccountName;
+  static const core::Property StorageAccountKey;
+  static const core::Property SASToken;
+  static const core::Property CommonStorageAccountEndpointSuffix;
+  static const core::Property ConnectionString;
+  static const core::Property Blob;
+  static const core::Property CreateContainer;
+
+  // Supported Relationships
+  static const core::Relationship Failure;
+  static const core::Relationship Success;
+
+  explicit PutAzureBlobStorage(std::string name, minifi::utils::Identifier uuid = minifi::utils::Identifier())
+    : PutAzureBlobStorage(name, uuid, nullptr) {
+  }
+
+  ~PutAzureBlobStorage() override = default;
+
+  bool supportsDynamicProperties() override { return true; }
+  void initialize() override;
+  void onSchedule(const std::shared_ptr<core::ProcessContext> &context, const std::shared_ptr<core::ProcessSessionFactory> &sessionFactory) override;
+  void onTrigger(const std::shared_ptr<core::ProcessContext> &context, const std::shared_ptr<core::ProcessSession> &session) override;
+
+  class ReadCallback : public InputStreamCallback {
+   public:
+    ReadCallback(uint64_t flow_size, azure::storage::BlobStorage& blob_storage_wrapper, const std::string &blob_name)
+      : flow_size_(flow_size)
+      , blob_storage_wrapper_(blob_storage_wrapper)
+      , blob_name_(blob_name) {
+    }
+
+    int64_t process(const std::shared_ptr<io::BaseStream>& stream) override {
+      std::vector<uint8_t> buffer;
+      buffer.reserve(flow_size_);
+      int read_ret = stream->read(buffer.data(), flow_size_);
```

Comment:
note that `std::vector<T>::data` is only guaranteed to return a pointer such that `[data(), data() + size())` is a valid range, and for empty vectors it might return whatever, we should use `resize` [1](https://stackoverflow.com/questions/38972443/use-stdvectordata-after-reserve)

Reply:
this is also an issue in `S3WrapperBase.cpp:147`, `PutS3Object.h:94` and `PublishMQTT.h:82`

Reply:
Fixed in [8853da](https://github.com/apache/nifi-minifi-cpp/pull/979/commits/8853da7346c55a08a326609e97ad09675443e7fb). Used read overload with vector parameter where resize is used already, where it was possible.